Drivers/Src/CANbus.c

```diff
@@ -1,75 +1,264 @@
 #include "CANbus.h"
 #include "BSP_CAN.h"
+#include "os.h"
 
-static void floatTo4Bytes(uint8_t f, uint8_t bytes[4]);
+/* Locking mechanism for the CAN bus.
+ * 
+ * The mutexes gate access to the BSP layer.
+ * 
+ * The mail sem4 counts the number of mailboxes
+ * in use, so that we can block if we can't get
+ * a hold of one.
+ * 
+ * The receive sem4 counts the number of messages
+ * currently in the receiving queue.
+ */
+static OS_MUTEX CANbus_TxMutex;
+static OS_MUTEX CANbus_RxMutex;
+static OS_SEM	CANbus_MailSem4;
+static OS_SEM	CANbus_ReceiveSem4;
+
+/**
+ * @brief   Releases hold of the mailbox semaphore.
+ * @note	Do not call directly.
+ */
+static void CANbus_Release(void) {
+	OS_ERR err;
+
+	OSSemPost(&CANbus_MailSem4,
+			  OS_OPT_POST_1,
+			  &err);
+}
+
+/**
+ * @brief	Increments the receive semaphore.
+ * @note	Do not call directly.
+ */
+static void CANbus_CountIncoming(void) {
+	OS_ERR err;
+
+	OSSemPost(&CANbus_ReceiveSem4,
+			  OS_OPT_POST_1,
+			  &err);
+}
 
 /**
  * @brief   Initializes the CAN system
  * @param   None
  * @return  None
  */
 void CANbus_Init(void) {
-    BSP_CAN_Init();
+	OS_ERR err;
+
+	OSMutexCreate(&CANbus_TxMutex,
+				  "CAN TX Lock",
+				  &err);
+
+	OSMutexCreate(&CANbus_RxMutex,
+				  "CAN RX Lock",
+				  &err);
+
+	OSSemCreate(&CANbus_MailSem4,
+                "CAN Mailbox Semaphore",
+                3,	// Number of mailboxes
+                &err);
+
+	OSSemCreate(&CANbus_ReceiveSem4,
+                "CAN Queue Counter Semaphore",
+                0,
+                &err);
+
+	// Initialize and pass interrupt hooks
+    BSP_CAN_Init(CANbus_CountIncoming, CANbus_Release);
 }
 
-/**
- * @brief   Transmits data onto the CANbus
- * @param   id : CAN id of the message
- * @param   payload : the data that will be sent.
- * @return  0 if data wasn't sent, otherwise it was sent.
- */
-int CANbus_Send(CANId_t id, CANPayload_t payload) {
-    uint8_t txdata[5];
+// Static method, call CANbus_Send or CANbus_BlockAndSend instead
+static int CANbus_SendMsg(CANId_t id, CANPayload_t payload) {
```

Comment:
Why does this return an `int` instead of `ErrorStatus` or `uint8_t` (return type of `BSP_CAN_Write()`. Also, the same thing for all of the following functions that return `int` as an error code